File: Painter's Partition/PaintersPartition.cpp
#include<bits/stdc++.h>
using namespace std;

int findFeasible(int boards[], int n, int limit){
    int sum=0, painters=1;
    for(int i=0; i<n; i++){
        sum+=boards[i];
        if(sum>limit){
            sum=boards[i];
            //we get a new painter and allocate to him
            painters++; 
        }
    }
    return painters;
}

int paintersPartition(int boards[], int n, int m){
    int totalLength=0, k=0;
    for(int i=0; i<n; i++){
        k=max(k, boards[i]);
        totalLength+=boards[i];
    }

    int low=k , high=totalLength;
    //here just we don't have to store the value in ans. the answer is already in low.
    while(low<high){
        int mid=(low+high)/2;
        int painters=findFeasible(boards, n, mid);
        if(painters<=m){
            high=mid;
        }
        else{
            low=mid+1;
        }
    }
    return low;
}

int main(){
    int arr[]={10, 20, 30, 40};
    int n=4;//number of boards
    int m=2;//number of painters
    cout<<"The minimum time to paint all boards is: "<<paintersPartition(arr, n, m);
    return 0;
}

File: Painter's Partition/PaintersPartition.py
'''
Solved using Binary Search in TC: O(n * log2(sum(boards)))
and constant space. 
'''
def findFeasible(boards, n, limit):
    sum = 0
    numberOfPainters = 1
    for i in range(n):
        sum += boards[i]
        if sum > limit:
            sum = boards[i]
            numberOfPainters += 1
    return numberOfPainters


def paintersPartition(boards, n, m):
    left = max(boards)
    right = sum(boards)
    result = -1
    while left <= right:
        mid = (left + right) // 2
        if findFeasible(boards, n, mid) <= m:
            result = mid
            right = mid - 1
        else:
            left = mid + 1
    return result

if __name__ == '__main__':
    arr=[10, 20, 30, 40]
    n=4
    #number of boards
    m=2
    #number of painters
    print(f'The minimum time to paint all boards is: {paintersPartition(arr, n, m)}')
